File: atcoder/other/idn_fb_c.cpp
#include <iostream>
#include <cmath>

using namespace std;

int pal[5010][5010];
int c[5010];
int dp[5010];


int main(){
  int n;
  char s[5010];
  cin >> n >> s;
  for (int i = 0; i < n; i++) cin >> c[i];
  
  for (int k = 0; k < n; k++) {
    for (int i = k, j = k; i >= 0 && j < n; i--, j++) {
      if (s[i] != s[j]) break;
      pal[i][j] = true;
    }
    for (int i = k, j = k+1; i >= 0 && j < n; i--, j++) {
      if (s[i] != s[j]) break;
      pal[i][j] = true;
    }
  }

  for (int i = 0; i < n+1; i++) dp[i] = n * (int)pow(10, 5);
  dp[0] = 0;
  for (int i = 0; i < n; i++) {
    for (int j = 0; j < i+1; j++) {
      if (pal[j][i]) dp[i+1] = min(dp[i+1], dp[j] + c[i-j]);
    }
  }
  cout << dp[n] << endl;
  return 0;
}

File: atcoder/other/idn_fb_c.py
n = int(input())
s = input()
c = list(map(int, input().split()))
pal = [[0 for _ in range(n)] for _ in range(n)]
for k in range(n):
    i, j = k, k
    while i >= 0 and j < n and s[i] == s[j]:
        pal[i][j] = True
        i -= 1
        j += 1
    i, j = k, k+1
    while i >= 0 and j < n and s[i] == s[j]:
        pal[i][j] = True
        i -= 1
        j += 1
dp = [(10 ** 5) * n for _ in range(n+1)]
dp[0] = 0
for i in range(n):
    for j in range(i+1):
        if pal[j][i]:
            dp[i+1] = min(dp[i+1], dp[j] + c[i-j])
print(dp[n])
